File: sdk/engine/src/engine.ts
import type {
  EngineConfig, NLUIEngine, Tool, Conversation, ChatOptions, EngineEvent, ToolSet, LLMConfig,
} from './types.js';
import { newAutoClient } from './llm/client.js';
import { ToolLoop } from './toolloop/loop.js';
import { GatewayCaller } from './gateway/caller.js';
import { buildTools } from './gateway/builder.js';
import { buildFromToolSet, parseToolSet } from './gateway/toolset.js';
import { loadSpec } from './gateway/openapi.js';
import { ConversationManager } from './conversation/manager.js';
import { MemoryStorage } from './conversation/storage.js';
import { buildSystemPrompt } from './prompt.js';
import type { Endpoint } from './types.js';

/**
 * Create an NLUIEngine instance.
 *
 * Loads all target specs, builds tools, constructs system prompt,
 * and wires up the LLM client + tool loop + conversation manager.
 */
export async function createEngine(config: EngineConfig): Promise<NLUIEngine> {
  // Mutable config state for runtime hot-reload
  let currentLLM: LLMConfig = { ...config.llm };
  let currentLang: 'zh' | 'en' | 'ja' = config.language ?? 'en';
  let currentProxy = config.proxy ?? '';

  let llmClient = newAutoClient(currentLLM);

  // Discover tools from all targets
  const allTools: Tool[] = [];
  const allEndpoints = new Map<string, Endpoint>();
  const targets = config.targets ?? [];

  for (const target of targets) {
    try {
      let tools: Tool[];
      let endpoints: Map<string, Endpoint>;

      if (target.tools) {
        const ts: ToolSet = typeof target.tools === 'string'
          ? parseToolSet(target.tools)
          : target.tools;
        ({ tools, endpoints } = buildFromToolSet(ts));
      } else if (target.spec) {
        const doc = await loadSpec(target.spec);
        const auth = target.auth ?? { type: '' };
        const baseURL = target.baseURL ?? '';
        ({ tools, endpoints } = buildTools(doc, target.name, baseURL, auth));
      } else {
        continue;
      }

      allTools.push(...tools);
      for (const [k, v] of endpoints) allEndpoints.set(k, v);
    } catch (err) {
      console.warn(`Skip target ${target.name}:`, err);
    }
  }

  // Build system prompt
  let systemPrompt = buildSystemPrompt(currentLang, targets, allTools);

  // Setup executor
  const caller = new GatewayCaller(allEndpoints);

  // Setup tool loop
  let loop = new ToolLoop(llmClient, caller);
  if (config.maxContextTokens) loop.setMaxContextTokens(config.maxContextTokens);
  if (config.confirm) loop.setConfirm(config.confirm);

  function rebuildLoop() {
    llmClient = newAutoClient(currentLLM);
    loop = new ToolLoop(llmClient, caller);
    if (config.maxContextTokens) loop.setMaxContextTokens(config.maxContextTokens);
    if (config.confirm) loop.setConfirm(config.confirm);
  }

  // Setup conversation manager
  const storage = config.storage ?? new MemoryStorage();
  const convMgr = new ConversationManager(storage);
  await convMgr.init();

  // Build the engine object
  const engine: NLUIEngine = {
    async chat(message: string, options?: ChatOptions): Promise<string> {
      const onEvent = buildEventRouter(options);

      let conv = options?.conversationId ? convMgr.get(options.conversationId) : null;
      const isNew = !conv;
      if (isNew) conv = convMgr.create('', systemPrompt);
      const convId = conv!.id;

      conv!.messages.push({ role: 'user', content: message });

      if (isNew) {
        const title = message.length > 30 ? message.slice(0, 30) + '...' : message;
        convMgr.updateTitle(convId, title);
      }

      const enabledTools = filterTools(allTools, conv!);

      try {
        const finalMessages = await loop.run(
          conv!.messages, enabledTools,
          options?.authToken ?? '',
          onEvent,
          options?.signal,
        );
        convMgr.updateMessages(convId, finalMessages);
        options?.onDone?.(convId);
      } catch (err) {
        options?.onError?.(err instanceof Error ? err : new Error(String(err)));
        // Still save whatever messages we have
        convMgr.updateMessages(convId, conv!.messages);
      }

      return convId;
    },

    createConversation(title?: string): Conversation {
      return convMgr.create(title ?? '', systemPrompt);
    },

    getConversation(id: string): Conversation | null {
      return convMgr.get(id);
    },

    listConversations(): Conversation[] {
      return convMgr.list();
    },

    deleteConversation(id: string): void {
      convMgr.delete(id);
    },

    editMessage(convId: string, index: number, content: string): void {
      convMgr.editMessage(convId, index, content);
    },

    deleteMessage(convId: string, index: number): void {
      convMgr.deleteMessage(convId, index);
    },

    deleteMessagesFrom(convId: string, index: number): void {
      convMgr.deleteMessagesFrom(convId, index);
    },

    async regenerateFrom(convId: string, index: number, options?: ChatOptions): Promise<void> {
      convMgr.deleteMessagesFrom(convId, index);
      const conv = convMgr.get(convId);
      if (!conv) throw new Error('Conversation not found');

      const enabledTools = filterTools(allTools, conv);
      const onEvent = buildEventRouter(options);

      const finalMessages = await loop.run(
        conv.messages, enabledTools,
        options?.authToken ?? '',
        onEvent,
        options?.signal,
      );
      convMgr.updateMessages(convId, finalMessages);
    },

    updateToolConfig(convId: string, enabledSources?: string[], disabledTools?: string[]): void {
      convMgr.updateToolConfig(convId, enabledSources, disabledTools);
    },

    tools(): Tool[] {
      return allTools;
    },

    getConfig() {
      return { llm: { ...currentLLM }, language: currentLang, proxy: currentProxy };
    },

    updateLLMConfig(apiBase: string, apiKey: string, model: string) {
      currentLLM = { ...currentLLM, apiBase, apiKey, model };
      rebuildLoop();
    },

    updateStream(enabled: boolean) {
      currentLLM = { ...currentLLM, stream: enabled };
      rebuildLoop();
    },

    updateLanguage(lang: 'zh' | 'en' | 'ja') {
      currentLang = lang;
      systemPrompt = buildSystemPrompt(lang, targets, allTools);
    },

    updateProxy(proxy: string) {
      currentProxy = proxy;
    },
  };

  return engine;
}

// ---- Internal helpers ----

function buildEventRouter(options?: ChatOptions): (event: EngineEvent) => void {
  return (event: EngineEvent) => {
    options?.onEvent?.(event);
    const data = event.data as Record<string, unknown>;
    switch (event.type) {
      case 'content_delta':
        options?.onDelta?.(data.delta as string);
        break;
      case 'tool_call':
        options?.onToolCall?.(data.name as string, data.arguments as string);
        break;
      case 'tool_result':
        options?.onToolResult?.(data.name as string, data.result as string);
        break;
    }
  };
}

function extractSource(toolName: string): string {
  const idx = toolName.indexOf('__');
  return idx > 0 ? toolName.slice(0, idx) : 'default';
}

function filterTools(tools: Tool[], conv: Conversation): Tool[] {
  if (!conv.enabledSources?.length && !conv.disabledTools?.length) return tools;

  return tools.filter(tool => {
    const name = tool.function.name;
    const source = extractSource(name);

    if (conv.enabledSources?.length && !conv.enabledSources.includes(source)) return false;
    if (conv.disabledTools?.includes(name)) return false;
    return true;
  });
}

File: engine/engine.go
package engine

import (
	"context"
	"fmt"
	"strings"

	"github.com/ZacharyZcR/NLUI/core/conversation"
	"github.com/ZacharyZcR/NLUI/core/llm"
	"github.com/ZacharyZcR/NLUI/core/toolloop"
)

// Type aliases — hosts only need to import engine.
type Event = toolloop.Event
type Executor = toolloop.Executor
type ConfirmFunc = toolloop.ConfirmFunc
type Tool = llm.Tool
type Message = llm.Message
type Conversation = conversation.Conversation

type Config struct {
	LLM          *llm.Client
	Executor     Executor
	Tools        []Tool
	SystemPrompt string
	MaxCtxTokens int
	ConvDir      string              // "" = in-memory only
	ConvMgr      *conversation.Manager // optional, reuse across reinit
}

type Engine struct {
	loop         *toolloop.Loop
	convMgr      *conversation.Manager
	tools        []Tool
	systemPrompt string
}

func New(cfg Config) *Engine {
	loop := toolloop.New(cfg.LLM, cfg.Executor)
	loop.SetMaxContextTokens(cfg.MaxCtxTokens)

	convMgr := cfg.ConvMgr
	if convMgr == nil {
		convMgr = conversation.NewManager(cfg.ConvDir)
	}

	return &Engine{
		loop:         loop,
		convMgr:      convMgr,
		tools:        cfg.Tools,
		systemPrompt: cfg.SystemPrompt,
	}
}

// Chat runs a full chat turn: get/create conversation → append user msg → loop → update messages.
// Returns the conversation ID used.
func (e *Engine) Chat(ctx context.Context, convID, message, authToken string, onEvent func(Event)) (string, error) {
	conv := e.convMgr.Get(convID)
	isNew := conv == nil
	if isNew {
		conv = e.convMgr.Create("", e.systemPrompt)
	}

	conv.Messages = append(conv.Messages, llm.Message{Role: "user", Content: message})

	if isNew {
		title := message
		if len([]rune(title)) > 30 {
			title = string([]rune(title)[:30]) + "..."
		}
		e.convMgr.UpdateTitle(conv.ID, title)
	}

	// Filter tools based on conversation config
	enabledTools := e.filterTools(conv)

	finalMessages, err := e.loop.Run(ctx, conv.Messages, enabledTools, authToken, onEvent)
	if err != nil {
		e.convMgr.UpdateMessages(conv.ID, finalMessages)
		return conv.ID, fmt.Errorf("chat: %w", err)
	}

	e.convMgr.UpdateMessages(conv.ID, finalMessages)
	return conv.ID, nil
}

func (e *Engine) SetConfirm(fn ConfirmFunc) {
	e.loop.SetConfirm(fn)
}

func (e *Engine) SetMaxContextTokens(n int) {
	e.loop.SetMaxContextTokens(n)
}

func (e *Engine) Tools() []Tool {
	return e.tools
}

func (e *Engine) SystemPrompt() string {
	return e.systemPrompt
}

func (e *Engine) CreateConversation(title string) *Conversation {
	return e.convMgr.Create(title, e.systemPrompt)
}

func (e *Engine) GetConversation(id string) *Conversation {
	return e.convMgr.Get(id)
}

func (e *Engine) ListConversations() []*Conversation {
	return e.convMgr.List()
}

func (e *Engine) DeleteConversation(id string) {
	e.convMgr.Delete(id)
}

// EditMessageAndRegenerate edits a message and regenerates from that point.
func (e *Engine) EditMessageAndRegenerate(ctx context.Context, convID string, msgIndex int, newContent, authToken string, onEvent func(Event)) error {
	if err := e.convMgr.EditMessage(convID, msgIndex, newContent); err != nil {
		return err
	}
	conv := e.convMgr.Get(convID)
	if conv == nil {
		return fmt.Errorf("conversation not found")
	}
	enabledTools := e.filterTools(conv)
	finalMessages, err := e.loop.Run(ctx, conv.Messages, enabledTools, authToken, onEvent)
	e.convMgr.UpdateMessages(convID, finalMessages)
	return err
}

// DeleteMessagesFrom deletes messages starting from the given index.
func (e *Engine) DeleteMessagesFrom(convID string, msgIndex int) error {
	return e.convMgr.DeleteMessagesFrom(convID, msgIndex)
}

// DeleteMessage deletes a single message at the given index.
func (e *Engine) DeleteMessage(convID string, msgIndex int) error {
	return e.convMgr.DeleteMessage(convID, msgIndex)
}

// UpdateToolConfig updates the tool configuration for a conversation.
func (e *Engine) UpdateToolConfig(convID string, enabledSources, disabledTools []string) error {
	return e.convMgr.UpdateToolConfig(convID, enabledSources, disabledTools)
}

// RegenerateFrom regenerates the conversation from a specific message index.
// Useful for retrying after the last assistant message.
func (e *Engine) RegenerateFrom(ctx context.Context, convID string, fromIndex int, authToken string, onEvent func(Event)) error {
	conv := e.convMgr.Get(convID)
	if conv == nil {
		return fmt.Errorf("conversation not found")
	}
	if fromIndex < 0 || fromIndex > len(conv.Messages) {
		return fmt.Errorf("invalid message index")
	}
	truncated := conv.Messages[:fromIndex]
	enabledTools := e.filterTools(conv)
	finalMessages, err := e.loop.Run(ctx, truncated, enabledTools, authToken, onEvent)
	e.convMgr.UpdateMessages(convID, finalMessages)
	return err
}

// filterTools returns the tools enabled for this conversation.
// If EnabledSources is empty, all tools are enabled.
// DisabledTools can block specific tools even within enabled sources.
func (e *Engine) filterTools(conv *Conversation) []Tool {
	if len(conv.EnabledSources) == 0 && len(conv.DisabledTools) == 0 {
		return e.tools // All tools enabled
	}

	var result []Tool
	for _, tool := range e.tools {
		toolName := tool.Function.Name
		source := extractSource(toolName)

		// Check if source is enabled (empty list = all enabled)
		if len(conv.EnabledSources) > 0 {
			sourceEnabled := false
			for _, s := range conv.EnabledSources {
				if s == source {
					sourceEnabled = true
					break
				}
			}
			if !sourceEnabled {
				continue
			}
		}

		// Check if tool is explicitly disabled
		toolDisabled := false
		for _, dt := range conv.DisabledTools {
			if dt == toolName {
				toolDisabled = true
				break
			}
		}
		if toolDisabled {
			continue
		}

		result = append(result, tool)
	}
	return result
}

// extractSource returns the source name from a tool name (e.g., "mcp__tool" -> "mcp")
func extractSource(toolName string) string {
	if idx := strings.Index(toolName, "__"); idx > 0 {
		return toolName[:idx]
	}
	return "default"
}
